File: src/solutions/2025/03/ocaml/main.ml
let options = Hashtbl.create 128;;

Hashtbl.add options "has_alternate" false;;
Hashtbl.add options "has_io" false;;

let solve (input : string) (_ : bool) (p2 : bool) : string =
  input
  |> String.split_on_char '\n'
  |> List.fold_left
       (fun acc line ->
          let res = ref 0 in
          let len = String.length line in
          let start = ref 0 in
          let max = if p2 then 12 else 2 in
          let bytes = Bytes.of_string line in
          for digit = 0 to max - 1 do
            let marked = ref 0 in
            let n = ref 0 in
            let t = max - 1 - digit in
            let l = len - t in
            for it = !start to l - 1 do
              let parsed = Bytes.get_int8 bytes it - 48 in
              if !n < parsed
              then (
                marked := it;
                n := parsed)
            done;
            start := !marked + 1;
            res := !res + (!n * int_of_float (10. ** float_of_int (max - 1 - digit)))
          done;
          acc + !res)
       0
  |> string_of_int
;;

let part_1 (input : string) (_is_test : bool) : string = solve input true false
let part_2 (input : string) (_is_test : bool) : string = solve input true true
let () = Aoc.run Sys.argv part_1 part_2 options

File: src/solutions/2025/03/c/main.c
#include "main.h"
#include "run.h"
#include "utils_num.h"
#include "utils_str.h"
#include <math.h>
#include <stdint.h>
#include <string.h>

static char *solve(const char *restrict input, const int32_t isTest,
                   const int32_t p2) {
   size_t sz;
   char **lines = str_splitc(input, '\n', &sz);
   int64_t res = 0;

   for (size_t i = 0; i < sz; i++) {
      char *line = lines[i];
      size_t len = strlen(line);
      size_t start = 0;
      size_t max = p2 ? 12 : 2;

      for (size_t digit = 0; digit < max; digit++) {
         size_t marked = 0;
         int64_t n = 0;
         size_t t = max - 1 - digit;
         size_t l = len - t;
         for (size_t it = start; it < l; it++) {
            int64_t parsed = line[it] - '0';
            if (n < parsed) {
               marked = it;
               n = parsed;
            }
         }

         start = marked + 1;
         res += n * (int64_t)pow(10, t);
      }

      free(line);
   }

   free(lines);
   return num_tostr(res);
}

static char *part1(const char *restrict input, const int32_t isTest) {
   return solve(input, isTest, 0);
}

static char *part2(const char *restrict input, const int32_t isTest) {
   return solve(input, isTest, 1);
}

int main(int argc, char *argv[]) {
   solution_options_t options = getoptions();
   return run(argc, argv, part1, part2, options);
}
